충남대_BE_김규현 Step2 관리자 페이지 코드 리뷰 요청

HappyGogildong opened the pull request:
Step2 코드 리뷰 요청입니다
merge된 코드를 끌어올 때 실수를 해서 이전 단계 커밋이 앞에 섞였습니다

HappyGogildong committed:
<feat>(api): 상품 조회 기능 추가

HappyGogildong committed:
<feat>(api/product): 상품 추가 기능 추가

HappyGogildong committed:
<feat, refactor>(/api/products): 관리자 페이지 상품 조회, 상품 추가 기능 구현

Thymeleaf 사용으로 RESTapi -> MVC로 변경함
상품 조회 메서드 수정 : "products.html" 사용
상품 추가 메서드 수정 : HTML 폼 사용

Breaks template에 products.html 추가
ProductController를 @RestController에서
@Controller로 변경
상품 추가 메서드를 Id, name, price, imageURL을 입력받게 변경

HappyGogildong committed:
<feat, refactor>(/api/products): 상품 수정 기능 구현

상품 수정 메서드 수정 : URL을 변경하여 수정 페이지로 이동하게 함.

Breaks 수정 페이지는 products.html의 상품 등록 폼을 재사용함(상품ID 입력란만 비활성화)
HTML 폼은 Patch요청을 지원하지 않기 때문에
GetMapping으로 변경함

HappyGogildong committed:
<feat, refactor>(/api/products): 상품 삭제 기능 구현

상품 삭제 메서드 수정 : URL을 변경, 리다이렉트 수행

HappyGogildong committed:
<refactor, chore>(/api/products): Validation 적용

상품 조회 메서드 변경 : 전체 상품 목록 조회용이므로 id를 받지 않게 변경
상품 수정, 삭제 메서드에 유효하지 않은 id 처리 로직 추가

brainbackdoor commented on README.md:
```diff
@@ -1,8 +1,8 @@
-[x] PATCH /api/product/{id} 상품 수정</br>
-[x] DELETE /api/product/{id} 상품 삭제
\ No newline at end of file
+### 2단계 관리자 화면
+---
+[x] 상품 조회
+[x] 상품 추가
+[x] 상품 수정
+[x] 상품 삭제
```
작은 기능을 구현할 때도 기능 요구사항을 상세하게 나눠보는 연습을 하시는 것이 좋습니다.
기능 요구사항을 작성하는 연습을 초기에 해두지 않으면, 나중에 복잡한 문제를 풀 때 어떻게 작업 단위를 나눠야 하는지 어려우실 수 있어요. 
무엇을 테스트해야 하는가? 하는 질문들을 종종 받는데요. 기능 요구사항으로 도출한 인수 조건들을 테스트하면 되거든요. 그러려면 평소에 좀 더 구현부가 상상될 정도로 기능 요구사항을 작성하는 연습을 해보는게 좋습니다.

HappyGogildong replied:
각 기능별로 요청 데이터, 응답 데이터를 적었습니다

brainbackdoor commented on src/main/java/gift/Application.java:
```diff
@@ -5,55 +5,93 @@
         SpringApplication.run(Application.class, args);
     }
 }
+@Component
+class ProductRepository {
+    final Map<Long, Product> products = new HashMap<>();
 
+    public Product get(long id) {
+        return products.get(id);
+    }
+
+    public boolean containsKey(long id) {
+        return products.containsKey(id);
+    }
+
+    public void add(Product product) {
+        products.put(product.id(), product);
+    }
+
+    public void delete(long id) {
+        products.remove(id);
+    }
+
+    public void update(long id, Product product) {
+        products.put(id, product);
+    }
+
+    public Collection<Product> findAll() {
+        return products.values();
+    }
+}
```
- 하나의 파일에 클래스를 모두 두는 이유가 있으실까요? 파일 분리를 해보시는건 어떨까요?

HappyGogildong replied:
repository, controller, entity를 분리했습니다

brainbackdoor commented on src/main/java/gift/Application.java:
```diff
@@ -5,55 +5,93 @@
+
+    public void delete(long id) {
+        products.remove(id);
+    }
+
+    public void update(long id, Product product) {
+        products.put(id, product);
+    }
+
+    public Collection<Product> findAll() {
+        return products.values();
+    }
+}
 
 record Product(long id, String name, int price, String imageURL){
 
 }
 
-@RestController
+@Controller
 class ProductController {
     //임시 저장소
-    private final Map<Long, Product> products = new HashMap<>();
-    private long currentId = 1L; //auto_increment용
+    private final ProductRepository products;
+
+    public ProductController(ProductRepository products) {
+        this.products = products;
+    }
 
     //상품 조회
-    @GetMapping("/product")
-    public ResponseEntity<Product> getProduct(@RequestParam(value = "id", defaultValue = "") long id) {
-        return products.containsKey(id) ?
-                new ResponseEntity<>(products.get(id), HttpStatus.OK) : new ResponseEntity<>(HttpStatus.NOT_FOUND);
+    @GetMapping("/api/products")
+    public String getProduct(Model model) {
+        model.addAttribute("allProducts",products.findAll());
+        return "products";
     }
```
- 음.. 기존의 기능을 유지한 상태에서 관리자 대시보드를 위한 API 가 추가되는 요구사항으로 이해했어요. 기능 요구사항을 보시면 이미 만든 상품 API를 이용하여 ajax 방식을 적용할 수 있다고도 표현한걸 보면, 기존 기능을 제거하고 새로운 요구사항이 도출되었다고 생각되지는 않네요

HappyGogildong replied:
단일 상품 조회 로직을 추가했습니다

brainbackdoor replied:
음.. 그런 의미가 아니었어요. 
1단계때 작성한 API는 그대로 두고, 어드민 API를 추가하는게 이번 단계의 기능 요구사항이에요. 
ajax로 기존 API를 활용할 수도 있겠구요.

brainbackdoor commented on src/main/java/gift/Application.java:
```diff
@@ -5,55 +5,93 @@
-        currentId++;
-        return new ResponseEntity<>(product, HttpStatus.CREATED);
+    //상품 추가,상품 수정
+    @PostMapping("/api/products")
+    public String createProduct(@RequestParam long id, @RequestParam String name, @RequestParam int price, @RequestParam String imageURL) {
```
- 왜 RequestBody로 받지 않고 RequestParam으로 받도록 작성하셨나요?

HappyGogildong replied:
HTML 폼 데이터는 url-encoded방식으로, key=value 형식으로 전달됩니다. 이 때 @RequestParam, @ModelAttribute를 사용하여 받을 수 있습니다
AJAX를 사용하여 JSON으로 변환하여 전달할경우, @RequestBody를 사용할 수 있습니다

brainbackdoor reviewed (changes requested):
빠르게 2단계 미션까지 진행해주셨네요! 
간단한 피드백 남겼으니 확인해보셔요~ 좋은 밤 보내세요! 🙏🏻

HappyGogildong committed:
<refactor>(/api/products):
피드백 내용 적용

repository, controller, entity, application 분리
API명세서에 맞게 Path parameter 사용한 단일 상품 조회 로직 추가
상품 수정 요청URL 변경

HappyGogildong committed:
<refactor>(/api/products):
피드백 내용 적용

repository, controller, entity, application 분리
API명세서에 맞게 Path parameter 사용한 단일 상품 조회 로직 추가
상품 수정 요청URL 변경

HappyGogildong commented:
PR이 열려있는 상태에서 push를 강제로 해버려서 conflict가 났었습니다..
아직 병합되지 않은 상태라 이전 커밋을 찾아 되돌리고 다시 push했습니다.

brainbackdoor commented on src/main/java/gift/ProductController.java:
```diff
@@ -0,0 +1,69 @@
+        return "redirect:/api/products";
+    }
+
+    //상품 추가 ModelAttribute 사용
+//    @PostMapping("/api/products")
+//    public String createProduct(@ModelAttribute Product product) {
+//        products.add(product);
+//        return "redirect:/api/products";
+//    }
```
- 사용하지 않는 코드는 그때그때 제거해두세요
- 불필요한 주석들은 제거해주세요.

brainbackdoor commented on src/main/java/gift/ProductController.java:
```diff
@@ -0,0 +1,69 @@
+//    }
+
+    //상품 수정(수정 페이지 이동)
+    @PostMapping("/api/products/{productId}")
+    public String updateProduct(@PathVariable long productId, @RequestParam String name, @RequestParam int price, @RequestParam String imageURL, RedirectAttributes redirectAttributes) {
+        if (products.get(productId) == null) {
+            redirectAttributes.addFlashAttribute("errorMessage", "상품이 존재하지 않음");
+            return "redirect:/api/products";}
```
- 포매팅은 습관적으로 해주세요.

brainbackdoor reviewed (approved):
Comment 하나 추가로 남겨두었는데요.
이번 요구사항은 기존 API를 없애고 관리자 대시보드를 작성하라는 의미가 아니에요. 
다음 단계 진행하면서 반영해주세요.
